File: app/src/main/java/com/lumiyaviewer/lumiya/react/OpportunisticExecutor.java
package com.lumiyaviewer.lumiya.react;

import com.lumiyaviewer.lumiya.Debug;
import java.util.HashSet;
import java.util.LinkedList;
import java.util.Queue;
import java.util.Set;
import java.util.concurrent.Executor;
import java.util.concurrent.locks.Condition;
import java.util.concurrent.locks.Lock;
import java.util.concurrent.locks.ReentrantLock;
import javax.annotation.Nonnull;

public class OpportunisticExecutor implements Executor {
    private final Lock lock = new ReentrantLock();
    private final Condition notEmpty = this.lock.newCondition();
    private final Queue<Runnable> queue = new LinkedList();
    private final Set<Runnable> runOnceRunnables = new HashSet();
    private final Thread thread;
    private final Runnable worker = new Runnable() {
        @Override
        public void run() {
            while (true) {
                try {
                    lock.lock();
                    
                    // Process regular queue items
                    Runnable task;
                    while ((task = queue.poll()) != null) {
                        lock.unlock();
                        try {
                            task.run();
                        } catch (Exception e) {
                            Debug.Warning(e);
                        }
                        lock.lock();
                    }
                    
                    // Process run-once runnables
                    boolean hasRunOnceWork = false;
                    if (!runOnceRunnables.isEmpty()) {
                        hasRunOnceWork = true;
                        
                        // Process all run-once runnables
                        var iterator = runOnceRunnables.iterator();
                        while (iterator.hasNext()) {
                            Runnable runOnceTask = iterator.next();
                            iterator.remove();
                            
                            lock.unlock();
                            try {
                                runOnceTask.run();
                            } catch (Exception e) {
                                Debug.Warning(e);
                            }
                            lock.lock();
                        }
                    }
                    
                    // If no work was done, wait for new tasks
                    if (!hasRunOnceWork) {
                        try {
                            notEmpty.await();
                        } catch (InterruptedException e) {
                            Thread.currentThread().interrupt();
                            break;
                        }
                    }
                    
                } catch (Exception e) {
                    Debug.Warning(e);
                } finally {
                    try {
                        lock.unlock();
                    } catch (Exception e) {
                        Debug.Warning(e);
                    }
                }
            }
        }
    };

    private class RunOnceExecutor implements Executor {
        private RunOnceExecutor() {
        }

        /* synthetic */ RunOnceExecutor(OpportunisticExecutor opportunisticExecutor, RunOnceExecutor runOnceExecutor) {
            this();
        }

        public void execute(@Nonnull Runnable runnable) {
            try {
                OpportunisticExecutor.this.lock.lock();
                OpportunisticExecutor.this.runOnceRunnables.add(runnable);
                OpportunisticExecutor.this.notEmpty.signalAll();
            } finally {
                OpportunisticExecutor.this.lock.unlock();
            }
        }
    }

    public OpportunisticExecutor(String str) {
        this.thread = new Thread(this.worker, str);
        this.thread.start();
    }

    public void execute(@Nonnull Runnable runnable) {
        try {
            this.lock.lock();
            if (Thread.currentThread().getId() == this.thread.getId() && this.queue.isEmpty()) {
                this.lock.unlock();
                runnable.run();
                this.lock.lock();
            } else {
                this.queue.offer(runnable);
                this.notEmpty.signalAll();
            }
        } catch (Throwable e) {
            Debug.Warning(e);
        } catch (Throwable th) {
            this.lock.unlock();
        }
        this.lock.unlock();
    }

    public RunOnceExecutor getRunOnceExecutor() {
        return new RunOnceExecutor(this, null);
    }
}

File: app/src/main/kotlin/com/lumiyaviewer/lumiya/react/OpportunisticExecutor.kt
package com.lumiyaviewer.lumiya.react

import com.lumiyaviewer.lumiya.Debug
import java.util.LinkedList
import java.util.Queue
import java.util.concurrent.Executor
import java.util.concurrent.locks.Condition
import java.util.concurrent.locks.ReentrantLock
import kotlin.concurrent.withLock

class OpportunisticExecutor(name: String) : Executor {
    private val lock = ReentrantLock()
    private val notEmpty: Condition = lock.newCondition()
    private val queue: Queue<Runnable> = LinkedList()
    private val runOnceRunnables = mutableSetOf<Runnable>()
    
    private val worker = Runnable {
        while (true) {
            try {
                lock.withLock {
                    // Process regular queue items
                    while (true) {
                        val task = queue.poll() ?: break
                        lock.unlock()
                        try {
                            task.run()
                        } catch (e: Exception) {
                            Debug.Warning(e)
                        }
                        lock.lock()
                    }
                    
                    // Process run-once runnables
                    var hasRunOnceWork = false
                    if (runOnceRunnables.isNotEmpty()) {
                        hasRunOnceWork = true
                        
                        val iterator = runOnceRunnables.iterator()
                        while (iterator.hasNext()) {
                            val runOnceTask = iterator.next()
                            iterator.remove()
                            
                            lock.unlock()
                            try {
                                runOnceTask.run()
                            } catch (e: Exception) {
                                Debug.Warning(e)
                            }
                            lock.lock()
                        }
                    }
                    
                    // If no work was done, wait for new tasks
                    if (!hasRunOnceWork) {
                        try {
                            notEmpty.await()
                        } catch (e: InterruptedException) {
                            Thread.currentThread().interrupt()
                            return@Runnable
                        }
                    }
                }
            } catch (e: Exception) {
                Debug.Warning(e)
            }
        }
    }
    
    private val thread = Thread(worker, name).apply { start() }
    
    inner class RunOnceExecutor : Executor {
        override fun execute(command: Runnable) {
            lock.withLock {
                runOnceRunnables.add(command)
                notEmpty.signalAll()
            }
        }
    }
    
    override fun execute(command: Runnable) {
        try {
            lock.lock()
            if (Thread.currentThread().id == thread.id && queue.isEmpty()) {
                lock.unlock()
                command.run()
                lock.lock()
            } else {
                queue.offer(command)
                notEmpty.signalAll()
            }
        } catch (e: Throwable) {
            Debug.Warning(e)
        } finally {
            try {
                lock.unlock()
            } catch (e: Exception) {
                // Ignore unlock errors
            }
        }
    }
    
    fun getRunOnceExecutor(): RunOnceExecutor = RunOnceExecutor()
}
